src/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs

```diff
@@ -36,37 +40,46 @@ public class CancellationTokenSource : IDisposable
         /// <summary><see cref="s_numPartitions"/> - 1, used to quickly mod into <see cref="_callbackPartitions"/>.</summary>
         private static readonly int s_numPartitionsMask = s_numPartitions - 1;
 
-        /// <summary>The current state of the CancellationTokenSource.</summary>
+        private volatile ManualResetEvent _kernelEvent; //lazily initialized if required.
+
+        // legal values for _state
+        private const int NotCanceledState = 1;
+        private const int NotifyingState = 2;
+        private const int NotifyingCompleteState = 3;
+
+        //_state uses the pattern "volatile int32 reads, with cmpxch writes" which is safe for updates and cannot suffer torn reads.
         private volatile int _state;
+
         /// <summary>The ID of the thread currently executing the main body of CTS.Cancel()</summary>
         /// <remarks>
         /// This helps us to know if a call to ctr.Dispose() is running 'within' a cancellation callback.
         /// This is updated as we move between the main thread calling cts.Cancel() and any syncContexts
         /// that are used to actually run the callbacks.
         /// </remarks>
         private volatile int _threadIDExecutingCallbacks = -1;
-        /// <summary>Tracks the running callback to assist ctr.Dispose() to wait for the target callback to complete.</summary>
+
+        private bool _disposed;
+
+        // we track the running callback to assist ctr.Dispose() to wait for the target callback to complete.
```

Comment:
CancellationTokenSource in coreclr is newer than the one in corert, and these fields were all reordered in coreclr... I'd prefer if we use the coreclr ordering / commenting / etc.